clippy_lints/src/unnecessary_self_imports.rs

```diff
@@ -0,0 +1,71 @@
+use clippy_utils::diagnostics::span_lint_and_then;
+use clippy_utils::source::snippet;
+use if_chain::if_chain;
+use rustc_ast::{Item, ItemKind, Path, PathSegment, UseTree, UseTreeKind};
+use rustc_errors::Applicability;
+use rustc_lint::{EarlyContext, EarlyLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::symbol::kw::SelfLower;
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for imports ending in `::{self};`.
+    ///
+    /// **Why is this bad?** In most cases, this can be written much more cleanly by omitting `self`.
+    ///
+    /// **Known problems:** Sometimes removing `::{self}` will break code (https://github.com/rust-lang/rustfmt/issues/3568).
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// use std::io::{self};
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// use std::io;
+    /// ```
+    pub UNNECESSARY_SELF_IMPORTS,
+    restriction,
+    "imports ending in `self`, which can be omitted"
+}
+
+declare_lint_pass!(UnnecessarySelfImports => [UNNECESSARY_SELF_IMPORTS]);
+
+impl EarlyLintPass for UnnecessarySelfImports {
+    fn check_item(&mut self, cx: &EarlyContext<'_>, item: &Item) {
+        if_chain! {
+            if let ItemKind::Use(
+                UseTree { kind: UseTreeKind::Nested(nodes), prefix: Path { span, .. }, .. }
+            ) = &item.kind;
```

Comment:
Just making `use_tree` available for a later suggestion...

```suggestion
            if let ItemKind::Use(use_tree) = &item.kind;
```